File: Baekjoon/yerim/정수 삼각형.cpp
// [1932] 정수 삼각형
#include <iostream>
#include <algorithm>
using namespace std;

int n, sum = 0;
int dp[505][505], triangle[505][505];

void input()
{
    cin >> n;
	for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= i; j++) {
			cin >> triangle[i][j];
		}
	}
}

void solve()
{   
    for (int i = 1; i <= n; i++) {
		for (int j = 1; j <= i; j++) {
			dp[i][j] = triangle[i][j] + max(dp[i - 1][j], dp[i - 1][j - 1]);
		}
	}

	for (int i = 1; i <= n; i++) {
		sum = max(sum, dp[n][i]);
	}

	cout << sum << endl;
}

int main(void)
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    input();
    solve();

    return 0;
}

File: Baekjoon/dongwan/정수 삼각형.py
n=int(input())
dp=[]
#dp에 싹 넣기 
for i in range(n):
    dp.append(list(map(int, input().split())))

#높이를 돌면서  
for i in range(1,n):
    #해당 행의 열들을 돌면서 
    for j in range(len(dp[i])):
        #처음건 비교할게 없이 위랑 더하기
        if j==0:
            dp[i][j]=dp[i][j]+dp[i-1][j]
        #마지막것도 비교할게 없이 위랑 더하기
        elif j==len(dp[i])-1: 
            dp[i][j]=dp[i][j]+dp[i-1][j-1]
        #아니라면 위에서 더 컸던거와 비교
        else:
            dp[i][j]=max(dp[i-1][j-1],dp[i-1][j])+dp[i][j]
print(max(dp[n-1]))